pandas/util/_decorators.py

```diff
@@ -272,23 +276,30 @@ def deprecate_nonkeyword_arguments(
 
     Parameters
     ----------
-    version : str, optional
-        The version in which positional arguments will become
-        keyword-only. If None, then the warning message won't
-        specify any particular version.
-
+    klass : Warning, optional
+        The warning class to use.
     allowed_args : list, optional
         In case of list, it must be the list of names of some
         first arguments of the decorated functions that are
         OK to be given as positional arguments. In case of None value,
         defaults to list of all arguments not having the
         default value.
-
     name : str, optional
         The specific name of the function to show in the warning
         message. If None, then the Qualified name of the function
         is used.
     """
+    from pandas.errors import (
+        Pandas4Warning,
+        Pandas5Warning,
+    )
+
+    if klass is Pandas4Warning:
+        version = "4.0"
+    elif klass is Pandas5Warning:
+        version = "5.0"
+    else:
```

Comment:
If you create the ABC `PandasWarning` as suggested above, then you could just do
```python
    if issubclass(klass, PandasWarning):
        version = klass.version()
    else:
        raise AssertionError(f"{type(klass)=} must be a versioned warning")
```